src/backend/joanie/lms_handler/backends/openedx.py

```diff
@@ -0,0 +1,126 @@
+"""
+Backend to connect Joanie to OpenEdX LMS
+"""
+import json
+import logging
+import re
+
+import requests
+from requests.auth import AuthBase
+
+from .base import BaseLMSBackend
+
+logger = logging.getLogger(__name__)
+
+
+class OpenEdXTokenAuth(AuthBase):
+    """Attach HTTP token authentication header to the given Request object."""
+
+    def __init__(self, token):
+        """Set-up token value in the instance."""
+        self.token = token
+
+    def __call__(self, request):
+        """Modify and return the request."""
+        request.headers.update(
+            {
+                "Content-Type": "application/json",
+                "X-Edx-Api-Key": self.token,
+            }
+        )
+
+        return request
+
+
+class TokenAPIClient(requests.Session):
+    """
+    A class `request.Session` that automatically authenticates against OpenEdX's preferred
+    authentication method up to Dogwood, given a secret token.
+
+    For more usage details, see documentation of the class `requests.Session` object:
+    https://requests.readthedocs.io/en/master/user/advanced/#session-objects
+    """
+
+    def __init__(self, token, *args, **kwargs):
+        """Extending the session object by setting the authentication token."""
+        super().__init__(*args, **kwargs)
+        self.auth = OpenEdXTokenAuth(token)
+
+
+class OpenEdXLMSBackend(BaseLMSBackend):
+    @property
+    def api_client(self):
+        """Instantiate and return an OpenEdX token API client."""
+        return TokenAPIClient(self.configuration["API_TOKEN"])
+
+    def extract_course_id(self, resource_link):
+        """Extract the LMS course id from the course run url."""
+        return re.match(self.configuration["COURSE_REGEX"], resource_link).group(
+            "course_id"
+        )
+
+    def parse_course_id(self, course_id):
+        """
+        Split an OpenEdX course id by organization, course and course run codes.
+
+        We first try splitting the course id as a version key
+        (course-v{version number}:{organization}+{course}+{run})
+        and fallback the old version
+        ({organization}/{course}/{run}).
+        """
+        course_id_pattern = re.compile(r"^course-v\d{1,2}:(?P<opaque_course_key>.*)$")
+        if match := course_id_pattern.match(course_id):
+            data = match.group("opaque_course_key").split("+")
+        else:
+            data = course_id.split("/")
+
+        data_length = len(data)
+
+        return (
+            data[0] if data_length > 0 else None, # Organization
+            data[1] if data_length > 1 else None, # Course number
```

Comment:
Are you sure this is a course number ? For me it's more a course name but maybe I'm wrong

Reply:
The `resource_link` used for test lets think that this is the course name but we take a look at real course_key, it seems to be more a number. I'll change my resource_link in tests.

`course-v1:ulg+108001+session09`
`course-v1:Agreenium+66005+session02`